File: Week_07/126.word-ladder-ii.go
/*
 * @lc app=leetcode.cn id=126 lang=golang
 *
 * [126] 单词接龙 II
 *
 * https://leetcode-cn.com/problems/word-ladder-ii/description/
 *
 * algorithms
 * Hard (38.47%)
 * Total Accepted:    24.9K
 * Total Submissions: 64.6K
 * Testcase Example:  '"hit"\n"cog"\n["hot","dot","dog","lot","log","cog"]'
 *
 * 给定两个单词（beginWord 和 endWord）和一个字典 wordList，找出所有从 beginWord 到 endWord
 * 的最短转换序列。转换需遵循如下规则：
 * 
 * 
 * 每次转换只能改变一个字母。
 * 转换后得到的单词必须是字典中的单词。
 * 
 * 
 * 说明:
 * 
 * 
 * 如果不存在这样的转换序列，返回一个空列表。
 * 所有单词具有相同的长度。
 * 所有单词只由小写字母组成。
 * 字典中不存在重复的单词。
 * 你可以假设 beginWord 和 endWord 是非空的，且二者不相同。
 * 
 * 
 * 示例 1:
 * 
 * 输入:
 * beginWord = "hit",
 * endWord = "cog",
 * wordList = ["hot","dot","dog","lot","log","cog"]
 * 
 * 输出:
 * [
 * ⁠ ["hit","hot","dot","dog","cog"],
 * ["hit","hot","lot","log","cog"]
 * ]
 * 
 * 
 * 示例 2:
 * 
 * 输入:
 * beginWord = "hit"
 * endWord = "cog"
 * wordList = ["hot","dot","dog","lot","log"]
 * 
 * 输出: []
 * 
 * 解释: endWord "cog" 不在字典中，所以不存在符合要求的转换序列。
 * 
 */
func appendWord(paths [][]string, word string) [][]string {
    result := make([][]string, 0, len(paths))
    for _, path := range paths {
        newpath := append([]string{}, path...)
        newpath = append(newpath, word)
        result = append(result, newpath)
    }
    return result
}
func findLadders(beginWord string, endWord string, wordList []string) [][]string {
    wordSet := make(map[string]bool)
    for _, str := range wordList {
        wordSet[str] = true
    }
    entry := map[string][][]string {
        beginWord: {{beginWord}},
    }
    for len(entry) > 0 {
        newentry := make(map[string][][]string)
        for word, paths := range entry {
            if word == endWord {
                return paths
            }
            wb := []byte(word)
            for i := 0; i < len(wb); i++ {
                for k := 0; k < 26; k++ {
                    c := 'a' + byte(k)
                    if c != wb[i] {
                        temp := append([]byte{}, wb...)
                        temp[i] = c
                        newword := string(temp)
                        if _, ok := wordSet[newword]; ok {
                            newpaths := appendWord(paths, newword)
                            if vpaths, ok := newentry[newword]; ok {
                                newentry[newword] = append(vpaths, newpaths...)
                            } else {
                                newentry[newword] = newpaths
                            }
                        }
                    }
                }
            }
        }
        entry = newentry
        for key, _ := range entry {
            delete(wordSet, key)
        }
    }
    return make([][]string, 0)
}

File: Week_07/126.word-ladder-ii.js
/*
 * @lc app=leetcode.cn id=126 lang=javascript
 *
 * [126] 单词接龙 II
 *
 * https://leetcode-cn.com/problems/word-ladder-ii/description/
 *
 * algorithms
 * Hard (38.47%)
 * Total Accepted:    24.9K
 * Total Submissions: 64.6K
 * Testcase Example:  '"hit"\n"cog"\n["hot","dot","dog","lot","log","cog"]'
 *
 * 给定两个单词（beginWord 和 endWord）和一个字典 wordList，找出所有从 beginWord 到 endWord
 * 的最短转换序列。转换需遵循如下规则：
 * 
 * 
 * 每次转换只能改变一个字母。
 * 转换后得到的单词必须是字典中的单词。
 * 
 * 
 * 说明:
 * 
 * 
 * 如果不存在这样的转换序列，返回一个空列表。
 * 所有单词具有相同的长度。
 * 所有单词只由小写字母组成。
 * 字典中不存在重复的单词。
 * 你可以假设 beginWord 和 endWord 是非空的，且二者不相同。
 * 
 * 
 * 示例 1:
 * 
 * 输入:
 * beginWord = "hit",
 * endWord = "cog",
 * wordList = ["hot","dot","dog","lot","log","cog"]
 * 
 * 输出:
 * [
 * ⁠ ["hit","hot","dot","dog","cog"],
 * ["hit","hot","lot","log","cog"]
 * ]
 * 
 * 
 * 示例 2:
 * 
 * 输入:
 * beginWord = "hit"
 * endWord = "cog"
 * wordList = ["hot","dot","dog","lot","log"]
 * 
 * 输出: []
 * 
 * 解释: endWord "cog" 不在字典中，所以不存在符合要求的转换序列。
 * 
 */
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {string[][]}
 */
var findLadders = function(beginWord, endWord, wordList) {
    let wordSet = new Set(wordList)
    if (!wordSet.has(endWord)) {
        return new Array()
    }
    let qmap = new Map()
    qmap.set(beginWord, [[beginWord]])
    while (qmap.size !== 0) {
        let nmap = new Map()
        for (let key of qmap.keys()) {
            let value = qmap.get(key)
            if (key === endWord) {
                return value
            }
            for (let i = 0; i < key.length; i++) {
                for (let j = 0; j < 26; j++) {
                    let c = String.fromCharCode('a'.charCodeAt(0) + j)
                    if (c === key[i]) continue
                    let word = key.substring(0, i) + c + key.substring(i + 1)
                    if (!wordSet.has(word)) continue
                    let newvalues = value.map(arr => [...arr, word])
                    if (!nmap.has(word)) {
                        nmap.set(word, [...newvalues])
                    } else {
                        nmap.get(word).push(...newvalues)
                    }
                }
            }
        }
        for (let w of nmap.keys()) {
            wordSet.delete(w)
        }
        qmap = nmap
    }
    return new Array()
};
